nav2_util/src/robot_utils.cpp

```diff
@@ -135,12 +121,60 @@ bool getTransform(
       rclcpp::get_logger("getTransform"),
       "Failed to get \"%s\"->\"%s\" frame transform: %s",
       source_frame_id.c_str(), target_frame_id.c_str(), ex.what());
-    return false;
+    return std::nullopt;
   }
+}
 
-  // Convert TransformStamped to TF2 transform
-  tf2::fromMsg(transform.transform, tf2_transform);
-  return true;
+template<>
+std::optional<tf2::Transform> getTransform(
```

Comment:
I would not - though - support this. In my opinion, consistency is important in large projects so folks understand what they're getting. If you're changing any of these to use `std::optional`, you must change them all and update them globally across the project

Reply:
Steve, would you mind clarifying the meaning of `this` (not the pointer :stuck_out_tongue: ) and "all`?

I do agree that constituency is important, but one needs to start somewhere anyway. It's hard to draw the line. If you are mentioning replacing all functions that return booleans instead of optional, then I'd agree that's the way to go, but that would be an immense PR and I didn't go that way since I first wanted to openly discuss what would be the preferred option.

Once I have directions from your side I'm happy to go full in, but I want to avoid spending hours changing the entire API to realize later that was a waste of time ;)

Reply:
This meaning to use optionals for the output of `getTransform` in `nav2_util`, given other methods in `robot_utils` are transforming data and not using that - which creates an inconsistent experience with users using the functions just with different input arguments.

So, its all or none. Either add your `getTransform` functions using the current styling or change all of the current versions and ones similar to it to use `std::optional` and update their local code usage. 

My preference would be that you just add the 1-2 new functions you need using the `bool` return, but if you did update the PR to make all of them use `std::optional` and update their in-line uses, I couldn't argue with that. I'll say that I try to keep the C++14 and newer feature uses to a minimum to keep it as approachable to the largest audience as possible. Sometimes its unavoidable and less approachable to _not_ use them, but this feels like a case that we're introducing C++17 features for the sake of it, rather than it being genuinely necessary. 

But, that's obviously just my single humble opinion. I wouldn't want to block you from contributing in the future over things like this, so if you want it I wouldn't block you from doing it as long as we maintain consistency. It just wouldn't be my recommendation.

Reply:
> I wouldn't want to block you from contributing in the future over things like this

Worry not Steve! I opened this PR to get a deeper understanding of how I can start contributing in the future without forcing you to review changes that are completely off of what's the vision for nav2.

As a summary of our discussion, this is what I plan to do now:

- [ ] merge quatri-plicated calls to the utility function to the `Source` class (discussion above)
- [ ] Remove the proposal about `std::optional`. Was just an approximation, I don't mind discarding it, I had fun doing it anyway. The main reason for me is that if we are in C++14, it's not worth introducing a hard dependency on 17. I can imagine an angry nacho trying to build this for a platform that's still not supporting c++17, and when understanding WHY we need 17 going bananas
- [ ] Use the more nav2/ROS 2 style of spitting booleans. I still don't like this at all... But I can't refactor the entire ROS 2 code base at this point.
- [ ] Add 2 extra util functions (THAT ARE NOT USED IN NAV2) but will extremely likely benefit users from the `nav2_util` package

Once you give me the OK I'll go ahead and make those changes.

PS: I still prefer this style of opening a PR and discussing rather than trying to write what I want to do in issues in terrible English. I prefer the code speak for me :P and I will never be bother to rollback a change I did

Reply:
Go ahead! I don’t love how that Tf set of calls is written either, but abstracting into a method would make it a lot less ugly. 

BTW: modern ROS 2 is on C++17, so you’re fine there. If there were going to have problems, it would be far before now.